Source/CesiumRuntime/Private/CesiumGeoreferenceComponent.cpp

```diff
@@ -135,90 +138,293 @@ void UCesiumGeoreferenceComponent::InaccurateMoveToECEF(
 }
 
+void UCesiumGeoreferenceComponent::OnComponentCreated() {
+  UE_LOG(
+      LogCesium,
+      Verbose,
+      TEXT("Called OnComponentCreated on component %s"),
+      *this->GetName());
+  Super::OnComponentCreated();
+  _initGeoreference();
+
+  // When the component is created, initialize its ECEF position with the
+  // position of the actor (but leave the rotation as it is).
+  // Do NOT use TransformUnrealToEcef, because it will assume
+  // that the given position is relative to the world origin
+  const glm::dvec3 absoluteLocation = _getAbsoluteLocationFromActor();
+  const glm::dmat4& unrealToEcef =
+      this->Georeference->getGeoTransforms()
+          .GetUnrealWorldToEllipsoidCenteredTransform();
+  const glm::dvec3 ecef = unrealToEcef * glm::dvec4(absoluteLocation, 1.0);
+  _setECEF(ecef, false);
+}
+
+void UCesiumGeoreferenceComponent::PostLoad() {
+  UE_LOG(
+      LogCesium,
+      Verbose,
+      TEXT("Called PostLoad on component %s"),
+      *this->GetName());
+  Super::PostLoad();
+  _initGeoreference();
+  this->_currentEcef = glm::dvec3(this->ECEF_X, this->ECEF_Y, this->ECEF_Z);
+}
+
+void UCesiumGeoreferenceComponent::_initGeoreference() {
+  if (!this->Georeference) {
+    this->Georeference = ACesiumGeoreference::GetDefaultGeoreference(this);
+  }
+  if (this->Georeference) {
+    UE_LOG(
+        LogCesium,
+        Verbose,
+        TEXT(
+            "Attaching CesiumGeoreferenceComponent callback to Georeference %s"),
+        *this->Georeference->GetFullName());
+    this->Georeference->OnGeoreferenceUpdated.AddUniqueDynamic(
+        this,
+        &UCesiumGeoreferenceComponent::HandleGeoreferenceUpdated);
+  }
+}
+
+void UCesiumGeoreferenceComponent::ApplyWorldOffset(
+    const FVector& InOffset,
+    bool bWorldShift) {
+  UE_LOG(
+      LogCesium,
+      Verbose,
+      TEXT("Called ApplyWorldOffset on component %s"),
+      *this->GetName());
+  UActorComponent::ApplyWorldOffset(InOffset, bWorldShift);
+  if (!this->FixTransformOnOriginRebase) {
+    return;
+  }
+
+  const UWorld* world = this->GetWorld();
+  if (!IsValid(world)) {
+    UE_LOG(
+        LogCesium,
+        Warning,
+        TEXT("CesiumGeoreferenceComponent %s is not spawned in world"),
+        *this->GetName());
+    return;
+  }
+  if (!IsValid(this->Georeference)) {
+    UE_LOG(
+        LogCesium,
+        Warning,
+        TEXT(
+            "CesiumGeoreferenceComponent %s does not have a valid Georeference"),
+        *this->GetName());
+    return;
+  }
+
+  // Compute the position that the world origin will have
+  // after the rebase, indeed by SUBTRACTING the offset
+  const glm::dvec3 oldWorldOriginLocation =
+      VecMath::createVector3D(world->OriginLocation);
+  const glm::dvec3 offset = VecMath::createVector3D(InOffset);
+  const glm::dvec3 newWorldOriginLocation = oldWorldOriginLocation - offset;
+
+  // Compute the absolute location based on the ECEF
+  // Do NOT use TransformEcefToUnreal, because it will return the
```

Comment:
I would be in favor of renaming `TransformEcefToUnreal` to `TransformEcefToUnrealRelative`, `GetEllipsoidCenteredToUnrealWorldTransform` to `GetEllipsoidCenteredToUnrealAbsoluteTransform` (the meaning of `World` is ambiguous), and so forth.